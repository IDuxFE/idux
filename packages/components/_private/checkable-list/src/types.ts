/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
