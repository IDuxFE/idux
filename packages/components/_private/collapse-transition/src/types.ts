/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
