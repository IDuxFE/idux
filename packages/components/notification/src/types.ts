/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, VNode } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'
import { ButtonProps } from '@idux/components/button'

// 挑出部分必填的属性
type PickRequire<T, U extends keyof T> = Pick<T, Exclude<keyof T, U>> & Required<Pick<T, U>>

export interface NotificationButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export interface SlotProps {
  visible?: boolean
  close?: () => void
}

export const notificationType = ['info', 'success', 'warning', 'error'] as const
export const notificationPlacement = ['topStart', 'topEnd', 'bottomStart', 'bottomEnd'] as const
export const notificationProps = {
  visible: IxPropTypes.bool,
  destroyOnHover: IxPropTypes.bool,
  duration: IxPropTypes.number,
  icon: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  closeIcon: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  type: IxPropTypes.oneOf<NotificationType>(['info', 'success', 'warning', 'error']),
  key: IxPropTypes.oneOfType<VKey>([String, Number, Symbol]),
  title: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  content: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  footer: IxPropTypes.oneOfType<string | NotificationButtonProps[] | VNode>([
    IxPropTypes.array(),
    IxPropTypes.vNode,
    String,
  ]),
  placement: IxPropTypes.oneOf<NotificationPlacement>(['topStart', 'topEnd', 'bottomStart', 'bottomEnd']),

  // event
  'onUpdate:visible': IxPropTypes.emit<(visible: boolean) => void>(),
  onClose: IxPropTypes.emit<(evt?: Event) => void>(),
}
export const notificationProviderProps = {
  offset: IxPropTypes.oneOfType<string | number | (number | string)[]>([String, Number, Array]),
  maxCount: IxPropTypes.number,
}

export type NotificationType = typeof notificationType[number]
export type NotificationPlacement = typeof notificationPlacement[number]
export type NotificationNodePropKey = 'title' | 'content' | 'footer'
export type NotificationProps = IxInnerPropTypes<typeof notificationProps>
export type NotificationPublicProps = IxPublicPropTypes<typeof notificationProps>
export type NotificationComponent = DefineComponent<
  Omit<HTMLAttributes, keyof NotificationPublicProps> & NotificationPublicProps
>
export type NotificationInstance = InstanceType<DefineComponent<NotificationProps>>

// 通过useNotification的配置
export interface NotificationOptions extends PickRequire<NotificationProps, 'title' | 'content'> {
  onDestroy?: (key: VKey) => void
}
export type NotificationPlacementMap = Record<NotificationPlacement, NotificationOptions[]>

export interface NotificationProviderRef {
  open: (options: NotificationOptions) => NotificationRef
  info: (options: Omit<NotificationOptions, 'type'>) => NotificationRef
  success: (options: Omit<NotificationOptions, 'type'>) => NotificationRef
  warning: (options: Omit<NotificationOptions, 'type'>) => NotificationRef
  error: (options: Omit<NotificationOptions, 'type'>) => NotificationRef
  update: (key: VKey, options: Partial<Omit<NotificationOptions, 'key'>>) => void
  destroy: (key: VKey | VKey[]) => void
  destroyAll: () => void
}

export interface NotificationRef {
  key: VKey
  update: (options: Partial<NotificationOptions>) => void
  destroy: () => void
}

export type NotificationProviderProps = IxInnerPropTypes<typeof notificationProviderProps>
export type NotificationProviderPublicProps = IxPublicPropTypes<typeof notificationProviderProps>
export type NotificationProviderComponent = DefineComponent<
  Omit<HTMLAttributes, keyof NotificationProviderPublicProps> & NotificationProviderPublicProps
>
export type NotificationProviderInstance = InstanceType<
  DefineComponent<NotificationProviderProps, NotificationProviderRef>
>
