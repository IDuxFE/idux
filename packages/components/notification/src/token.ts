/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { NotificationProviderRef } from './types'
import type { InjectionKey } from 'vue'

export const notificationProviderToken: InjectionKey<NotificationProviderRef> = Symbol('notificationProviderToken')
