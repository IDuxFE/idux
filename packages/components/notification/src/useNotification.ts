/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { NotificationProviderRef } from './types'

import { inject } from 'vue'

import { throwError } from '@idux/cdk/utils'

import { notificationProviderToken } from './token'

export const useNotification = (): NotificationProviderRef => {
  const notificationProviderRef = inject(notificationProviderToken, null)
  if (notificationProviderRef === null) {
    return throwError('components/notification', '<IxNotificationProvider> not found.')
  }

  return notificationProviderRef
}
