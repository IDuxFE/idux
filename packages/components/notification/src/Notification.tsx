/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { NotificationNodePropKey, NotificationProps, SlotProps } from './types'
import type { NotificationConfig } from '@idux/components/config'
import type { ComputedRef, Slots } from 'vue'

import { computed, defineComponent, onBeforeUnmount, onMounted, watchEffect } from 'vue'

import { isArray, isString } from 'lodash-es'

import { callEmit, useControlledProp } from '@idux/cdk/utils'
import { IxButton } from '@idux/components/button'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'
import { IxSpace } from '@idux/components/space'

import { notificationProps } from './types'

const defaultCloseIcon = 'close'

const defaultIcon = {
  success: 'check-circle-filled',
  error: 'close-circle-filled',
  info: 'info-circle-filled',
  warning: 'exclamation-circle-filled',
} as const

export default defineComponent({
  name: 'IxNotification',
  props: notificationProps,
  setup(props, { slots }) {
    const commonCfg = useGlobalConfig('common')
    const comPrefix = computed(() => `${commonCfg.prefixCls}-notification`)
    const config = useGlobalConfig('notification')
    const wrapCls = useClasses(props, comPrefix)
    const icon = useIcon(props, config)
    const closeIcon = useCloseIcon(props, config)
    const { visible, close, onMouseEnter, onMouseLeave } = useVisible(props, config)

    return () => {
      const iconNode = icon.value && getIconNode(icon.value)
      const closeIconNode = getIconNode(closeIcon.value)
      return (
        <div v-show={visible.value}>
          <div class={wrapCls.value} onMouseenter={onMouseEnter} onMouseleave={onMouseLeave}>
            <div v-show={icon.value} class={`${comPrefix.value}-icon`}>
              {iconNode}
            </div>
            <span class={`${comPrefix.value}-close-icon`} onClick={close}>
              {closeIconNode}
            </span>
            <div class={`${comPrefix.value}-main`}>
              <div class={`${comPrefix.value}-title`}>{getNode(props, slots, 'title')}</div>
              <div class={`${comPrefix.value}-content`}>{getNode(props, slots, 'content')}</div>
            </div>
            <div class={`${comPrefix.value}-footer`}>
              {getNode(props, slots, 'footer', {
                visible: visible.value,
                close,
              })}
            </div>
          </div>
        </div>
      )
    }
  },
})

function useClasses(props: NotificationProps, comPrefix: ComputedRef<string>) {
  return computed(() => [comPrefix.value, { [`${comPrefix.value}-${props.type}`]: props.type }])
}

function useIcon(props: NotificationProps, config: NotificationConfig) {
  return computed(() => {
    const iconMap = { ...defaultIcon, ...config.icon }
    return props.icon ?? (props.type && iconMap[props.type])
  })
}

function useCloseIcon(props: NotificationProps, config: NotificationConfig) {
  return computed(() => props.closeIcon ?? config.closeIcon ?? defaultCloseIcon)
}

function useVisible(props: NotificationProps, config: NotificationConfig) {
  const duration = computed(() => props.duration ?? config.duration)
  const destroyOnHover = computed(() => props.destroyOnHover ?? config.destroyOnHover)
  const autoClose = computed(() => duration.value > 0)

  const [visible, setVisible] = useControlledProp(props, 'visible', false)

  let timer: number | null = null

  const startTimer = () => {
    timer = setTimeout(() => close(), duration.value)
  }

  const clearTimer = () => {
    if (timer) {
      clearTimeout(timer)
      timer = null
    }
  }

  const close = () => {
    clearTimer()
    setVisible(false)
    callEmit(props.onClose)
  }

  const onMouseEnter = () => {
    if (autoClose.value && !destroyOnHover.value) {
      clearTimer()
    }
  }
  const onMouseLeave = () => {
    if (autoClose.value && !destroyOnHover.value) {
      startTimer()
    }
  }

  onMounted(() => {
    watchEffect(() => {
      clearTimer()
      if (visible.value && autoClose.value) {
        startTimer()
      }
    })
  })

  onBeforeUnmount(() => clearTimer())

  return { visible, close, onMouseEnter, onMouseLeave }
}

function getIconNode(icon: NotificationProps['icon']) {
  return isString(icon) ? <IxIcon name={icon}></IxIcon> : icon
}

function getNode<T extends NotificationProps, K extends NotificationNodePropKey>(
  props: T,
  slots: Slots,
  nodeKey: K,
  slotProps?: SlotProps,
) {
  const slotMap = {
    title: 'title',
    content: 'default',
    footer: 'footer',
  } as const
  const getHandlePropsMap = {
    title: (propsValue: NotificationProps[K]) => propsValue,
    content: (propsValue: NotificationProps[K]) => propsValue,
    footer: (propsValue: NotificationProps[K]) => getFooterNode(propsValue),
  } as const
  return slots[slotMap[nodeKey]]?.(slotProps) ?? getHandlePropsMap[nodeKey]?.(props[nodeKey])
}

function getFooterNode(footer?: NotificationProps['footer']) {
  if (!footer) {
    return ''
  }
  if (!isArray(footer)) {
    return footer
  }
  const footerButtons = footer.map(item => {
    const { text, ...rest } = item
    return <IxButton {...rest}>{text}</IxButton>
  })
  return <IxSpace>{footerButtons}</IxSpace>
}
