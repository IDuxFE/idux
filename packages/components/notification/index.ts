/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { NotificationComponent } from './src/types'

import Notification from './src/Notification'
import NotificationProvider from './src/NotificationProvider'

const IxNotification = Notification as unknown as NotificationComponent
const IxNotificationProvider = NotificationProvider as unknown as NotificationComponent

export { IxNotification, IxNotificationProvider }
export { useNotification } from './src/useNotification'

export type {
  NotificationInstance,
  NotificationPublicProps as NotificationProps,
  NotificationType,
  NotificationPlacement,
  NotificationRef,
  NotificationProviderRef,
} from './src/types'
