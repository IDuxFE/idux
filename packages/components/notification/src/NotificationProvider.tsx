/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type {
  NotificationOptions,
  NotificationPlacement,
  NotificationPlacementMap,
  NotificationProviderProps,
  NotificationRef,
} from './types'
import type { CommonConfig, NotificationConfig } from '@idux/components/config'

import { ComputedRef, Ref, TransitionGroup, cloneVNode, computed, defineComponent, isVNode, provide, ref } from 'vue'

import { isArray, isUndefined, pickBy } from 'lodash-es'

import { CdkPortal } from '@idux/cdk/portal'
import { Logger, VKey, callEmit, convertArray, convertCssPixel, uniqueId } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { usePortalTarget } from '@idux/components/utils'

import Notification from './Notification'
import { notificationProviderToken } from './token'
import { notificationPlacement, notificationProviderProps, notificationType } from './types'

const groupPositions = {
  topStart: ['top', 'left'],
  topEnd: ['top', 'right'],
  bottomStart: ['bottom', 'left'],
  bottomEnd: ['bottom', 'right'],
} as const

export default defineComponent({
  name: 'IxNotificationProvider',
  inheritAttrs: false,
  props: notificationProviderProps,
  setup(props, { slots, expose }) {
    const commonCfg = useGlobalConfig('common')
    const config = useGlobalConfig('notification')
    const mergedPrefixCls = computed(() => `${commonCfg.prefixCls}-notification`)
    const mergedPortalTarget = usePortalTarget(props, config, commonCfg, mergedPrefixCls)

    const maxCount = computed(() => props.maxCount ?? config.maxCount)
    const { notifications, loadContainer, open, info, success, warning, error, update, destroy, destroyAll } =
      useNotification(maxCount)
    const apis = { open, info, success, warning, error, update, destroy, destroyAll }
    const placementNotifications = usePlacementNotifications(props, config, notifications)

    provide(notificationProviderToken, apis)
    expose(apis)

    if (__DEV__) {
      if (props.target) {
        Logger.warn('components/notification', 'the `target` was deprecated, please use `container` instead.')
      }
    }

    return () => {
      const getChild = (notifications: NotificationOptions[]) => {
        return notifications.map(item => {
          const { key, visible = true, content, contentProps, ...restProps } = item
          const onClose = () => destroy(key!)
          const mergedProps = { key, visible, onClose }
          const contentNode = isVNode(content) ? cloneVNode(content, contentProps, true) : content
          return (
            <Notification {...mergedProps} {...restProps}>
              {contentNode}
            </Notification>
          )
        })
      }

      const placementGroup = Object.entries(placementNotifications.value).map(([key, notificationsRecord]) => {
        const placement = key as NotificationPlacement
        const child = getChild(notificationsRecord)
        const moveName = getMoveName(placement, commonCfg)
        const position = getGroupPosition(props, config, placement)
        return (
          <TransitionGroup
            key={placement}
            tag="div"
            appear
            name={moveName}
            class={[`${mergedPrefixCls.value}-wrapper`, `${mergedPrefixCls.value}-wrapper-${placement}`]}
            style={position}
          >
            {child}
          </TransitionGroup>
        )
      })

      return (
        <>
          {slots.default?.()}
          <CdkPortal target={mergedPortalTarget.value} load={loadContainer.value}>
            {placementGroup}
          </CdkPortal>
        </>
      )
    }
  },
})

function usePlacementNotifications(
  props: NotificationProviderProps,
  config: NotificationConfig,
  notifications: Ref<NotificationOptions[]>,
) {
  return computed(() => {
    const initValue = getPlacementInitValue<NotificationOptions[]>(() => []) as NotificationPlacementMap
    const { destroyOnHover, duration, closeIcon, placement } = config
    if (notifications.value.length === 0) {
      return initValue
    }
    return notifications.value.reduce((acc, item) => {
      const notificationProps = getRealObj({
        destroyOnHover,
        duration,
        closeIcon,
        placement,
        ...item,
      }) as unknown as NotificationOptions
      const curPlacement = notificationProps.placement!
      acc[curPlacement].push(notificationProps)
      return acc
    }, initValue)
  })
}

function useNotification(maxCount: ComputedRef<number>) {
  const notifications = ref<NotificationOptions[]>([])

  const getCurrIndex = (key: VKey) => {
    return notifications.value.findIndex(notification => notification.key === key)
  }

  const add = (item: NotificationOptions) => {
    const currIndex = item.key ? getCurrIndex(item.key) : -1
    if (currIndex !== -1) {
      notifications.value.splice(currIndex, 1, item)
      return item.key!
    }

    if (notifications.value.length >= maxCount.value) {
      notifications.value = notifications.value.slice(-maxCount.value + 1)
    }

    const key = item.key ?? uniqueId('ix-notification')
    notifications.value.push({ ...item, key })
    return key
  }

  const update = (key: VKey, item: Partial<NotificationOptions>) => {
    const currIndex = getCurrIndex(key)
    if (currIndex !== -1) {
      const newItem = { ...notifications.value[currIndex], ...item }
      notifications.value.splice(currIndex, 1, newItem)
    }
  }

  const destroy = (key: VKey | VKey[]) => {
    const keys = convertArray(key)
    keys.forEach(key => {
      const currIndex = getCurrIndex(key)
      if (currIndex !== -1) {
        const item = notifications.value.splice(currIndex, 1)
        callEmit(item[0].onDestroy, key)
      }
    })
  }

  const destroyAll = () => {
    notifications.value = []
  }

  const loadContainer = ref(false)

  const open = (options: NotificationOptions): NotificationRef => {
    const key = add(options)
    const ref = {
      key,
      update: (options: Partial<NotificationOptions>) => update(key, options),
      destroy: () => destroy(key),
    }
    if (!loadContainer.value) {
      loadContainer.value = true
    }
    return ref
  }

  const [info, success, warning, error] = notificationType.map(
    type => (options: Omit<NotificationOptions, 'type'>) => open({ ...options, type }),
  )

  return { notifications, loadContainer, open, info, success, warning, error, update, destroy, destroyAll }
}

function getGroupPosition(
  props: NotificationProviderProps,
  config: NotificationConfig,
  placement: NotificationPlacement,
) {
  const realPlacement = placement ?? config.placement
  const realOffset = props.offset ?? config.offset
  return getPosition(realOffset, realPlacement)
}

function getPosition(offset: NotificationProviderProps['offset'], placement: NotificationPlacement) {
  const offsets = isArray(offset) ? offset : [offset, offset]
  const [verticalOffset, horizontalOffset] = offsets.map(convertCssPixel)
  const [verticalPosition, horizontalPosition] = groupPositions[placement]
  return {
    [verticalPosition]: verticalOffset,
    [horizontalPosition]: horizontalOffset,
  }
}

function getMoveName(placement: NotificationPlacement, commonCfg: CommonConfig) {
  const prefixCls = commonCfg.prefixCls
  const moveNameMap = {
    topEnd: `${prefixCls}-move-end`,
    bottomEnd: `${prefixCls}-move-end`,
    topStart: `${prefixCls}-move-start`,
    bottomStart: `${prefixCls}-move-start`,
  } as const
  return moveNameMap[placement]
}

function getRealObj<T>(obj: Record<string, T>) {
  return pickBy<T>(obj, (val: T) => !isUndefined(val))
}

function getPlacementInitValue<T>(initValueFn: () => T) {
  return notificationPlacement.reduce((acc, item) => {
    const cloneValue = initValueFn()
    return {
      ...acc,
      [item]: cloneValue,
    }
  }, {} as Record<NotificationPlacement, T>)
}
