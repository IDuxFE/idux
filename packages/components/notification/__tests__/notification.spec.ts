import { MountingOptions, flushPromises, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork, wait } from '@tests'

import { IxButton } from '@idux/components/button'
import { IxIcon } from '@idux/components/icon'

import Notification from '../src/Notification'
import { NotificationProps } from '../src/types'

const notificationWrapCls = 'ix-notification'

describe('Notification', () => {
  const titleText = 'This is a title'
  const contentText = 'This is a notification'

  const NotificationMount = (options?: MountingOptions<Partial<NotificationProps>>) => {
    const { props, ...rest } = options || {}
    const _options = {
      props: {
        duration: 100,
        visible: true,
        title: titleText,
        content: contentText,
        ...props,
      },
      attachTo: 'body',
      ...rest,
    } as MountingOptions<NotificationProps>
    return mount(Notification, { ..._options })
  }

  renderWork<NotificationProps>(Notification, {
    props: { visible: true },
    slots: {
      title: () => titleText,
      default: () => contentText,
    },
  })

  test('visible work', async () => {
    const onUpdateVisible = jest.fn()
    const wrapper = NotificationMount({
      props: {
        visible: false,
        'onUpdate:visible': onUpdateVisible,
      },
    })
    await flushPromises()

    expect(wrapper.isVisible()).toBe(false)

    await wrapper.setProps({ visible: true })

    expect(wrapper.isVisible()).toBe(true)

    await wait(100)

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('destroyOnHover work', async () => {
    const onUpdateVisible = jest.fn()
    const wrapper = NotificationMount({
      propsData: {
        visible: false,
        'onUpdate:visible': onUpdateVisible,
      },
    })
    const wrapperDom = wrapper.find(`.${notificationWrapCls}`)

    await wrapper.setProps({
      visible: true,
    })
    await flushPromises()
    await wrapperDom.trigger('mouseenter')
    await wait(100)

    expect(onUpdateVisible).not.toBeCalled()

    await wrapperDom.trigger('mouseleave')
    await wait(100)

    expect(onUpdateVisible).toBeCalledWith(false)

    await wrapper.setProps({
      visible: true,
      destroyOnHover: true,
    })
    await wrapperDom.trigger('mouseenter')
    await wait(100)

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('duration work', async () => {
    const onUpdateVisible = jest.fn()
    const wrapper = NotificationMount({
      props: {
        visible: true,
        'onUpdate:visible': onUpdateVisible,
      },
    })

    await wait(90)

    expect(onUpdateVisible).not.toBeCalled()

    await wait(10)

    expect(onUpdateVisible).toBeCalledWith(false)

    await wrapper.setProps({
      visible: true,
      duration: 200,
    })
    await wait(190)

    expect(onUpdateVisible).toHaveBeenCalledTimes(1)

    await wait(10)

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('icon work', async () => {
    const wrapper = NotificationMount()
    await flushPromises()

    expect(wrapper.find(`.${notificationWrapCls}-icon`).isVisible()).toBe(false)

    await wrapper.setProps({ icon: 'up' })

    expect(wrapper.find('.ix-icon-up').exists()).toBe(true)

    await wrapper.setProps({ icon: h(IxIcon, { name: 'down' }) })

    expect(wrapper.find('.ix-icon-down').exists()).toBe(true)
  })

  test('closeIcon work', async () => {
    const wrapper = NotificationMount()
    await flushPromises()

    expect(wrapper.find('.ix-icon-close').exists()).toBe(true)

    await wrapper.setProps({
      closeIcon: 'right',
    })

    expect(wrapper.find('.ix-icon-close').exists()).toBe(false)
    expect(wrapper.find('.ix-icon-right').exists()).toBe(true)

    await wrapper.setProps({
      closeIcon: h(IxIcon, { name: 'left' }),
    })

    expect(wrapper.find('.ix-icon-right').exists()).toBe(false)
    expect(wrapper.find('.ix-icon-left').exists()).toBe(true)
  })

  test('type work', async () => {
    const wrapper = NotificationMount({
      props: {
        type: 'info',
      },
    })
    await flushPromises()

    expect(wrapper.find(`.${notificationWrapCls}`).classes()).toContain(`${notificationWrapCls}-info`)
    expect(wrapper.find('.ix-icon-info-circle-filled').exists()).toBe(true)
  })

  test('title work', async () => {
    const newTitle = 'new title'
    const wrapper = NotificationMount({
      props: {
        title: newTitle,
      },
    })
    await flushPromises()

    expect(wrapper.find(`.${notificationWrapCls}-title`).text()).toBe(newTitle)

    const newSlotTitle = 'new slot title'
    const newWrapper = NotificationMount({
      props: {
        title: newTitle,
      },
      slots: {
        title: newSlotTitle,
      },
    })
    await flushPromises()

    expect(newWrapper.find(`.${notificationWrapCls}-title`).text()).toBe(newSlotTitle)
  })

  test('content work', async () => {
    const newContent = 'This is new content'
    const wrapper = NotificationMount({
      props: {
        content: newContent,
      },
    })
    await flushPromises()

    expect(wrapper.find(`.${notificationWrapCls}-content`).text()).toBe(newContent)

    const newSlotContent = 'new slot content'
    const newWrapper = NotificationMount({
      props: {
        content: newSlotContent,
      },
      slots: {
        title: newSlotContent,
      },
    })
    await flushPromises()

    expect(newWrapper.find(`.${notificationWrapCls}-content`).text()).toBe(newSlotContent)
  })

  test('footer work', async () => {
    const wrapper = NotificationMount()
    await flushPromises()

    expect(wrapper.find(`.${notificationWrapCls}-footer`).text()).toBe('')

    const footerStr = 'footerStr'
    await wrapper.setProps({
      footer: footerStr,
    })

    expect(wrapper.find(`.${notificationWrapCls}-footer`).text()).toContain(footerStr)

    const footerVNode = h('div', { id: 'footVNode' }, footerStr)
    await wrapper.setProps({
      footer: footerVNode,
    })

    expect(wrapper.find('#footVNode').exists()).toBe(true)

    const clickFn = jest.fn()
    await wrapper.setProps({
      footer: [
        {
          mode: 'primary',
          text: 'ok',
          onClick: clickFn,
        },
      ],
    })
    expect(wrapper.find(`.${notificationWrapCls}-footer`).findComponent(IxButton).exists()).toBe(true)
    expect(wrapper.findComponent(IxButton).text()).toContain('ok')

    await wrapper.findComponent(IxButton).trigger('click')

    expect(clickFn).toBeCalled()
  })
})
