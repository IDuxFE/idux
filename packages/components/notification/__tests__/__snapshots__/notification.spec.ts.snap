// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification render work 1`] = `
"<div>
  <div class=\\"ix-notification\\">
    <div class=\\"ix-notification-icon\\" style=\\"display: none;\\">
      <!---->
    </div><span class=\\"ix-notification-close-icon\\"><i class=\\"ix-icon ix-icon-close\\" role=\\"img\\" aria-label=\\"close\\"></i></span>
    <div class=\\"ix-notification-main\\">
      <div class=\\"ix-notification-title\\">This is a title</div>
      <div class=\\"ix-notification-content\\">This is a notification</div>
    </div>
    <div class=\\"ix-notification-footer\\"></div>
  </div>
</div>"
`;
